ldcontext/builder_new_test.go

```diff
@@ -0,0 +1,524 @@
+			assert.Len(t, c.privateAttrs, 0)
+		})
+	}
+
+	t.Run("_meta", func(t *testing.T) {
+		b := makeKindBuilder()
+		assert.False(t, b.TrySetValue("_meta", ldvalue.String("hi")))
+		assert.Equal(t, 0, b.Build().attributes.Count())
+	})
+}
+
+func TestKindBuilderAttributesCopyOnWrite(t *testing.T) {
+	value1, value2 := ldvalue.String("value1"), ldvalue.String("value2")
+
+	b := makeKindBuilder().SetValue("attr", value1)
+
+	c1 := b.Build()
+	jsonhelpers.AssertEqual(t, value1, c1.attributes.Get("attr"))
+
+	b.SetValue("attr", value2)
+
+	c2 := b.Build()
+	jsonhelpers.AssertEqual(t, value2, c2.attributes.Get("attr"))
+	jsonhelpers.AssertEqual(t, value1, c1.attributes.Get("attr")) // unchanged
+}
+
+func TestKindBuilderPrivate(t *testing.T) {
+	expectPrivateRefsToBe := func(t *testing.T, c Context, expectedRefs ...ldattr.Ref) {
+		if assert.Equal(t, len(expectedRefs), c.PrivateAttributeCount()) {
+			for i, expectedRef := range expectedRefs {
+				a, ok := c.PrivateAttributeByIndex(i)
+				assert.True(t, ok)
+				assert.Equal(t, expectedRef, a)
+			}
+			_, ok := c.PrivateAttributeByIndex(len(expectedRefs))
+			assert.False(t, ok)
+		}
+		_, ok := c.PrivateAttributeByIndex(-1)
+		assert.False(t, ok)
+	}
+
+	t.Run("using Refs", func(t *testing.T) {
+		attrRef1, attrRef2, attrRef3 := ldattr.NewRef("a"), ldattr.NewRef("/b/c"), ldattr.NewRef("d")
+		c := makeKindBuilder().
+			PrivateRef(attrRef1, attrRef2).PrivateRef(attrRef3).
+			Build()
+
+		expectPrivateRefsToBe(t, c, attrRef1, attrRef2, attrRef3)
+	})
+
+	t.Run("using strings", func(t *testing.T) {
+		s1, s2, s3 := "a", "/b/c", "d"
+		b0 := makeKindBuilder().
+			PrivateRef(ldattr.NewRef(s1), ldattr.NewRef(s2)).PrivateRef(ldattr.NewRef(s3))
+		b1 := makeKindBuilder().
+			Private(s1, s2, s3)
+		assert.Equal(t, b0, b1)
+	})
+
+	t.Run("RemovePrivate", func(t *testing.T) {
+		b := makeKindBuilder().Private("a", "/b/c", "d", "/b/c")
+		b.RemovePrivate("/b/c")
+		c := b.Build()
+
+		expectPrivateRefsToBe(t, c, ldattr.NewRef("a"), ldattr.NewRef("d"))
+	})
+
+	t.Run("RemovePrivateRef", func(t *testing.T) {
+		b := makeKindBuilder().Private("a", "/b/c", "d", "/b/c")
+		b.RemovePrivateRef(ldattr.NewRef("/b/c"))
+		c := b.Build()
+
+		expectPrivateRefsToBe(t, c, ldattr.NewRef("a"), ldattr.NewRef("d"))
+	})
+
+	t.Run("copy on write", func(t *testing.T) {
+		b0 := makeKindBuilder().Private("a")
+
+		c0 := b0.Build()
+		expectPrivateRefsToBe(t, c0, ldattr.NewRef("a"))
+
+		b0.Private("b")
+		c1 := b0.Build()
+		expectPrivateRefsToBe(t, c1, ldattr.NewRef("a"), ldattr.NewRef("b"))
+		expectPrivateRefsToBe(t, c0, ldattr.NewRef("a")) // unchanged
+
+		b0.RemovePrivateRef(ldattr.NewRef("a"))
+		c2 := b0.Build()
+		expectPrivateRefsToBe(t, c2, ldattr.NewRef("b"))
+		expectPrivateRefsToBe(t, c1, ldattr.NewRef("a"), ldattr.NewRef("b")) // unchanged
+		expectPrivateRefsToBe(t, c0, ldattr.NewRef("a"))                     // unchanged
+	})
+}
+
+// MULTI KIND TESTS (loosely adapted from builder_multi_test.go)
```

Comment:
Tests from here down are newer / more handcrafted by me, because the multi-kind logic of ContextBuilder is somewhat different than what MultiBuilder can do.